src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -91,117 +95,189 @@ export class KernelExecution implements IDisposable {
         if (!editor) {
             return;
         }
-        const cancelTokenSource = new MultiCancellationTokenSource();
-        this.documentExecutions.set(document, cancelTokenSource);
+        this.documentExecutions.set(document, new CancellationTokenSource());
 
         traceInfo('Update notebook execution state as running');
         await editor.edit((edit) =>
             edit.replaceMetadata({ ...document.metadata, runState: vscodeNotebookEnums.NotebookRunState.Running })
         );
-        const codeCellsToExecute = document.cells
+        document.cells
             .filter((cell) => cell.cellKind === vscodeNotebookEnums.CellKind.Code)
-            .filter((cell) => cell.document.getText().trim().length > 0)
-            .map((cell) => {
-                const cellExecution = this.executionFactory.create(cell, isPythonKernelConnection(this.metadata));
-                this.cellExecutions.set(cellExecution.cell, cellExecution);
-                return cellExecution;
-            });
-        cancelTokenSource.token.onCancellationRequested(
-            () => codeCellsToExecute.forEach((cell) => cell.cancel()),
-            this,
-            this.disposables
-        );
+            .forEach((cell) => this.createCellExecution(cell));
 
         try {
-            const notebook = await notebookPromise;
-            const kernel = this.getKernel(document);
-            codeCellsToExecute.forEach((exec) => traceCellMessage(exec.cell, 'Ready to execute'));
```

Comment:
Moved all of this code into a new method, so that running a cell or all cells & queueing other cells end up in the same place - chained promises & in the right order.